plugin-server/src/main/ingestion-queues/session-recordings-consumer.ts

```diff
@@ -0,0 +1,138 @@
+
+export const startSessionRecordingEventsConsumer = async ({
+    teamManager,
+    kafka,
+    producer,
+}: {
+    teamManager: TeamManager
+    kafka: Kafka
+    producer: KafkaProducerWrapper
+}) => {
+    /*
+        For Session Recordings we need to prepare the data for ClickHouse.
+        Additionally, we process `$performance_event` events which are closely
+        tied to session recording events.
+
+        NOTE: it may be safer to also separate processing of
+        `$performance_event` but for now we'll keep it in the same consumer.
+    */
+
+    const consumer = kafka.consumer({ groupId: 'session-recordings' })
+    setupEventHandlers(consumer)
+
+    status.info('🔁', 'Starting session recordings consumer')
+
+    const eachBatch: EachBatchHandler = async ({ batch, heartbeat }) => {
+        status.debug('🔁', 'Processing batch', { size: batch.messages.length })
+        for (const message of batch.messages) {
+            if (!message.value) {
+                status.warn('⚠️', 'invalid_message', {
+                    reason: 'empty',
+                    partition: batch.partition,
+                    offset: message.offset,
+                })
+                await producer.queueMessage({ topic: KAFKA_SESSION_RECORDING_EVENTS_DLQ, messages: [message] })
+                continue
+            }
+
+            let messagePayload: RawEventMessage
+            let event: PipelineEvent
+
+            try {
+                // NOTE: we need to parse the JSON for these events because we
+                // need to add in the team_id to events, as it is possible due
+                // to a drive to remove postgres dependency on the the capture
+                // endpoint we may only have `token`.
+                messagePayload = JSON.parse(message.value.toString())
+                event = JSON.parse(messagePayload.data)
+            } catch (error) {
+                status.warn('⚠️', 'invalid_message', {
+                    reason: 'invalid_json',
+                    error: error,
+                    partition: batch.partition,
+                    offset: message.offset,
+                })
+                await producer.queueMessage({ topic: KAFKA_SESSION_RECORDING_EVENTS_DLQ, messages: [message] })
+                continue
+            }
+
+            status.debug('⬆️', 'processing_session_recording', { uuid: messagePayload.uuid })
+
+            if (!messagePayload.team_id && !event.token) {
+                status.warn('⚠️', 'invalid_message', {
+                    reason: 'no_token',
+                    partition: batch.partition,
+                    offset: message.offset,
+                })
+                await producer.queueMessage({ topic: KAFKA_SESSION_RECORDING_EVENTS_DLQ, messages: [message] })
+                continue
+            }
+
+            const teamId =
+                messagePayload.team_id ?? (event.token ? await teamManager.getTeamIdByToken(event.token) : null)
```

Comment:
question: Keep in mind JS/TS is not my forte:

We are checking for falsy `event.token` and `messagePayload.team_id` and skipping the message if that's the case.

However, it seems that immediately afterwards we handle both of these cases: if `messagePayload.team_id` is falsy, we use `event.token` to try to fetch it, and if `event.token` is also falsy then we set to `null`. Given that this double falsy case should never happen (as the `!messagePayload.team_id && !event.token` condition would be true), wouldn't the ternary operator be unnecessary? I.e. if `messagePayload.team_id` is null, then `event.token` cannot be null, thus can be used by `getTeamIdByToken` safely.

Reply:
I'll refactor this to be clearer. iirc it's like this to make the type checking happy but is a little convoluted.

Reply:
Oh yeah, I imagined the extra checks could be due to the type-checker, thanks for the explanation.